File: typescript/src/logging.ts
import { PingStats } from "./stats.js";

export interface StatsLogger {
  showCurrentStats(stats: PingStats): void;
  showStatsSummary(stats: PingStats): void;
}

export class ConsoleStatsLogger implements StatsLogger {
  showCurrentStats(stats: PingStats) {
    const successPercentage = Math.round(stats.successRate() * 100);
    const message = `SUCCESS RATE : ${successPercentage}%`;

    console.log(message);
  }

  showStatsSummary(stats: PingStats) {
    const successPercentage = Math.round(stats.successRate() * 100);
    const nbTries = stats.total();
    const message = `Successfuly pinged ${successPercentage}% of the requests over ${nbTries} tries.`;

    console.log(message);
  }
}

File: rust/src/logging.rs
use crate::stats::PingStats;

pub trait StatsLogger {
    fn show_current_stats(&self, stats: &PingStats);
    fn show_stats_summary(&self, stats: &PingStats);
}

pub struct ConsoleStatsLogger {}

impl StatsLogger for ConsoleStatsLogger {
    fn show_current_stats(&self, stats: &PingStats) {
        let success_percentage = (stats.success_rate() * 100.0).round() as u16;
        println!("SUCCESS RATE : {success_percentage}%");
    }

    fn show_stats_summary(&self, stats: &PingStats) {
        let success_percentage = (stats.success_rate() * 100.0).round() as u16;
        let nb_tries = stats.total();

        println!("Successfuly pinged {success_percentage}% of the requests over {nb_tries} tries.")
    }
}
